wandb/apis/public.py

```diff
@@ -1916,22 +1916,30 @@ def file(self, name):
     def upload_file(self, path, root="."):
         """
         Arguments:
-            path (str): name of file to upload.
-            root (str): the root path to save the file relative to.  i.e.
-                If you want to have the file saved in the run as "my_dir/file.txt"
-                and you're currently in "my_dir" you would set root to "../"
+            path (str): local path to the file to upload.
+            root (str): the root path to save the file in wandb.
 
         Returns:
             A `File` matching the name argument.
         """
+        path = os.path.normpath(os.path.join(os.getcwd(), os.path.relpath(path, os.getcwd())))
+        if not os.path.isfile(path):
+            raise Exception(f"Either the file {path} doesn't exist or it's a directory!")
+        path_ext = os.path.splitext(path)[1]
+        if path_ext == "":
+            raise Exception(f"File {path} must contain a file extension, otherwise it won't be uploaded!")
+        if root == "/":
+            raise Exception("'/' is not a valid directory, to use the root directory use '.' instead.")
+        root_ext = os.path.splitext(root)[1]
+        if root_ext not in [path_ext, ""]:
+            raise Exception(f"Local file extension ({path_ext}) doesn't match target file extension ({root_ext})!")
+        name = os.path.normpath(f"{root}/{os.path.basename(path)}" if root_ext == "" else root)
```

Comment:
Could you help me understand why you think this is necessary? If the intent is to allow setting arbitrary destination path similar to, say, `cp`, why would you want to force the extension to be the same? Also, the file need not have an extension in the first place.

Reply:
So from my understanding, the extension should be the same as it doesn't make too much sense to rename `config.yaml` to `config.json`, as that can be misleading. This is why in the case of specifying the actual destination file path, it should have the same extension (or None if it's a directory). So this basically allows you to rename the file.